lib/log_displayer.js

```diff
@@ -6,61 +6,73 @@ let url = require('url')
 let liner = require('../lib/line_transform')
 const colorize = require('./colorize')
 
-function readLogs (logplexURL) {
-  let u = url.parse(logplexURL)
-  if (u.query && u.query.includes('srv')) {
-    return readLogsV1(logplexURL)
-  } else {
-    return readLogsV2(logplexURL)
+async function logDisplayer (heroku, options) {
+  function readLogs (logplexURL) {
+    let u = url.parse(logplexURL)
+    if (u.query && u.query.includes('srv')) {
+      return readLogsV1(logplexURL)
+    } else {
+      return readLogsV2(logplexURL)
+    }
   }
-}
 
-function readLogsV1 (logplexURL) {
-  return new Promise(function (resolve, reject) {
-    let res = cli.got.stream(logplexURL)
-    res.setEncoding('utf8')
-    liner.setEncoding('utf8')
-    res.pipe(liner)
-    liner.on('data', line => cli.log(colorize(line)))
-    res.on('end', resolve)
-    res.on('error', reject)
-  })
-}
-
-function readLogsV2 (logplexURL) {
-  return new Promise(function (resolve, reject) {
-    const u = url.parse(logplexURL, true)
-    const isTail = u.query.tail && u.query.tail === 'true'
-    const userAgent = process.env.HEROKU_DEBUG_USER_AGENT || 'heroku-run'
-    const proxy = process.env.https_proxy || process.env.HTTPS_PROXY
-    const es = new EventSource(logplexURL, {
-      proxy,
-      headers: {
-        'User-Agent': userAgent
-      }
+  function readLogsV1 (logplexURL) {
+    return new Promise(function (resolve, reject) {
+      let res = cli.got.stream(logplexURL)
+      res.setEncoding('utf8')
+      liner.setEncoding('utf8')
+      res.pipe(liner)
+      liner.on('data', line => cli.log(colorize(line)))
+      res.on('end', resolve)
+      res.on('error', reject)
     })
+  }
 
-    es.onerror = function (err) {
-      if (!isTail) {
-        resolve()
-        es.close()
-      }
+  function readLogsV2 (logplexURL) {
+    return new Promise(function (resolve, reject) {
+      const u = url.parse(logplexURL, true)
+      const isTail = u.query.tail && u.query.tail === 'true'
+      const userAgent = process.env.HEROKU_DEBUG_USER_AGENT || 'heroku-run'
+      const proxy = process.env.https_proxy || process.env.HTTPS_PROXY
+      const es = new EventSource(logplexURL, {
+        proxy,
+        headers: {
+          'User-Agent': userAgent
+        }
+      })
+      let lines = 0
 
-      if (err && (err.status === 404 || err.status === 403)) {
-        reject(new Error('Log stream timed out. Please try again.'))
-        es.close()
+      es.onerror = function (err) {
+        if (!isTail) {
+          if (lines === 0) {
+            // if app emits too many logs logplex will not return historical logs
+            // in this case, tail instead
+            logDisplayer(heroku, {
```

Comment:
The message would probably get lost in log scroll, but is it worth informing the user of what's happening here?

Reply:
I considered that. It would have to be on stderr but there isn't anything wrong with that. I'm leaning towards no though.

For apps that are spewing so many logs that they aren't being stored in redis I don't think makes any practical difference to the user and would be extra noise. If you piped it to grep or something it would show up and be pretty annoying I think. It also doesn't matter if the logs are historical since they're only going to be likely a few milliseconds different. The overhead the CLI takes to start would make a bigger difference in what is actually displayed.

Now for apps that simply have had 0 logs in the last 7 days, I think there is a stronger case. This is why I want a boolean flag to come from the API so I can tell the difference.

Though there is also the (maybe impossible) edge case to consider where an app IS flagged into the non-redis bucket BUT also doesn't have any logs in the last 7 days.

I think for all of these though the right thing is to switch to tail and let it fill up until the number of logs lines requested. Timing out maybe if it doesn't happen in time (maybe a relatively short timeout) and THEN showing a warning letting the user know what's going on.

Reply:
writing this out made me realize a probable bug in this PR. I don't think I'm accounting for `heroku logs` when you don't specify `-n`